zeebe/atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java

```diff
@@ -923,11 +935,32 @@ public long getFirstCommitIndex() {
    *
    * @param firstCommitIndex The first commit index.
    */
-  public void setFirstCommitIndex(final long firstCommitIndex) {
+  public void setFirstCommitIndex(final long firstCommitIndex, final long lastFlushedIndex) {
     if (this.firstCommitIndex == 0) {
       if (firstCommitIndex == 0) {
         return;
       }
+
+      // To detect if the current leader has experienced data loss we need to check both its
+      // commitIndex and the lastFlushedIndex:
+      // the previous leader may have already committed lastFlushedIndex before crashing,
+      // but failed to replicate the commit to the new leader.
+      // In this situation, the commitIndex of the leader may be smaller than the one in this node,
+      // but both nodes agree that events <= lastFlushedIndex have been persisted.
+      if (firstCommitIndex < commitIndex && lastFlushedIndex < commitIndex) {
+        final var errorMessage =
+            String.format(
+                """
+                   Expected to set first commit position to after restart, but firstCommitIndex(%d), lastflushedIndex(%d) < commitIndex(%d). \
+                   While commitIndex is the last committed index, persisted in the metadata file. \
+                   This means a majority of nodes has lost committed data: \
+                   This node will become inactive to avoid overwriting previously committed data, \
+                   but the leader have formed a quorum and will continue to commit new events, \
+                   creating an inconsistent timeline of events: \
+                   THE CLUSTER SHOULD BE STOPPED IMMEDIATELY to further prevent inconsistencies.""",
+                firstCommitIndex, lastFlushedIndex, commitIndex);
+        throw new RuntimeException(errorMessage);
```

Comment:
:wrench: Let's make that an `IllegalStateException`, just like in the other case.

Reply:
Probably `IllegalArgumentException` is better as the problems comes from the argument. I will change it in `RaftLog` as well.

Reply:
I sort of disagree because there's no one providing illegal arguments except ourselves. The system is in an illegal state overall, that the data is coming in via arguments is just accidental. It would be different if this was coming straight from user input.